src/avram/model.cr

```diff
@@ -116,12 +116,12 @@ abstract class Avram::Model
         ensure_value_is_string!({{value_generator}}.call)
       end
 
-      def self.ensure_value_is_string(value : String)
+      def self.ensure_value_is_string!(value : String)
         value
       end
 
-      def self.ensure_value_is_string(value)
-        {% raise "Your primary_key value_generator proc does not return a string" %}
+      def self.ensure_value_is_string!(value)
+        raise "Your primary_key value_generator proc must return a string"
```

Comment:
Yeah I'm still having troubles with this, not yet happy with the solution.

As a compile-time macro, it always raises the exception, not sure why.

I was experimenting with an alternative idea, where if the primary key is a String, you're forced to implement a method in your class with the signature `def generate_id : String`, but so far I'm failing to make that work 🤔

Reply:
Try doing 

```crystal
def self.ensure_value_is_string!(value : String)
  "good job"
end
def self.ensure_value_is_string!(value : Nil)
  {% raise "can't be nil" %}
end
def self.ensure_value_is_string!(value : T) forall T
  {% raise "can't be T" %}
end
```

That's how we're using it in Lucky https://github.com/luckyframework/lucky/blob/3de0f33544ea99f4e67a874badc345b3b2a08fc0/src/lucky/renderable.cr#L134-L168 I wonder if it needs that type in there for the catch all? If that doesn't work, then just leave it runtime and we can come back to it later.

Reply:
Seems like it's ending up in the `Nil` case:

```
In spec/support/models/discount.cr:5:5

 5 | primary_key id : String, value_generator: ->{ Random::Secure.hex }
     ^----------
Error: Your primary_key value_generator proc must return a string, but it's returning Nil
```

This is weird, [`Random::Hex` returns a string](https://crystal-lang.org/api/1.11.2/Random.html#hex%28n%3AInt%3D16%29%3AString-instance-method)

Reply:
hmm :thinking:  I wonder if `{{ value_generator }}` has to be like resolved, or have `.id` called on it or something because it's a Proc being passed in to a macro. It might not be a `Proc`, but maybe like `ProcLiteral` in macro sense or something weird like that.

Reply:
I found a different solution, let's see if you like it:

- Instead of passing a proc to the `primary_key` method, when the type is `String`, it'll include a module `Avram::StringPrimaryKeyMethods`
- That module contains an abstract function `abstract def generate_id : String`
- If your class doesn't then have a method called `generate_id` this is the error:

```
In spec/support/models/discount.cr:1:1

 1 | class Discount < BaseModel
     ^
Error: abstract `def Avram::StringPrimaryKeyMethods#generate_id()` must be implemented by Discount
```

What do you think?

Reply:
Oh, that's not a bad idea. I like it :+1:

Reply:
Ha! That doesn't work: `generate_id` needs to be a class method so we can invoke it before instantiating the class, but there is no way to define abstract class methods.

(the type-checking sorcery you suggested didn't work, I'll try something else now)

Reply:
Can't figure this out.

Anyways, I realized there is no need to force passing a proc. Simply passing a method call will work as well.
This then leads to a much better compile error on it's own, even if I would have loved a custom one:

```
There was a problem expanding macro 'primary_key'

Code in spec/support/models/discount.cr:5:5

 5 | primary_key id : String, Random.rand
     ^
Called macro defined in src/avram/model.cr:92:3

 92 | macro primary_key(type_declaration, value_generator = nil)

Which expanded to:

 > 11 |
 > 12 |
 > 13 |       def self.primary_key_value_generator : String
                                                     ^-----
Error: method Discount.primary_key_value_generator must return String but it is returning Float64
```

Reply:
Hmm. What about just having a value defined on the TypeDeclaration like other columns? 

```
column status : Int32 = 4

primary_key id : String = BaseModel.generate_id
```

Then you know you can check things on the type_dec.value. 

I'm not sure the best way to solve this issue though. Pretty tricky problem :thinking:

Reply:
Big fan, that syntax feels way more natural and allow passing a proc without the need to type the argument key.

The error is basically the same:

```
There was a problem expanding macro 'primary_key'

Code in spec/support/models/discount.cr:5:5

 5 | primary_key id : String = Random.rand
     ^
Called macro defined in src/avram/model.cr:92:3

 92 | macro primary_key(type_declaration)

Which expanded to:

 > 14 |
 > 15 |
 > 16 |       def self.primary_key_value_generator : String
                                                     ^-----
Error: method Discount.primary_key_value_generator must return String but it is returning Float64
```

Reply:
That's strange the compiler thinks it might return a Float :confused:  I'll try to take a look in a bit if I can get a little more free time. This is great progress though!

Reply:
In that example I changed the code to make it fail. `Random.rand` returns a float.

The actual functioning code in the tests uses `Random::Secure.hex` which returns a string.
I did that to illustrate what compile error the user might get